Add Support for HTTP

pratikpandey21 opened the pull request.

pratikpandey21 commented:
@arpitbbhayani @JyotinderSingh Review this once, so that we can merge and have others start on HTTP tasks

JyotinderSingh commented:
> @arpitbbhayani @JyotinderSingh Review this once, so that we can merge and have others start on HTTP tasks

Looks like there's some conflicts

pratikpandey21 commented:
> > @arpitbbhayani @JyotinderSingh Review this once, so that we can merge and have others start on HTTP tasks
> 
> Looks like there's some conflicts

Fixed.

JyotinderSingh commented on internal/server/httpServer.go:
```diff
@@ -0,0 +1,127 @@
+}
+
+// NewHTTPServer
+// TODO: This isn't ideal as we create a separate instance of ShardManager, which we want to be 1, which all servers share
```
Is this comment outdated? Aren't we already creating and sharing a common shard manager across both servers?

pratikpandey21 replied:
Yes, correct. This is outdated.

JyotinderSingh commented on internal/shard/shard_thread.go:
```diff
@@ -110,7 +110,12 @@ func (shard *ShardThread) executeCommand(op *ops.StoreOp) []byte {
 
-	// The following commands could be handled at the server level, however, we can randomly let any shard handle them
+	// Till the time we refactor to handle QWATCH differently using HTTP Streaming/SSE
+	if op.HTTPOp {
+		return diceCmd.Eval(op.Cmd.Args, shard.store)
+	}
```
Won't this cause a segfault in case the user ends up triggering QWATCH or QUNWATCH?

pratikpandey21 replied:
Yes, there is a separate issue to handle QWATCH etc with HTTP. Once that is added, this can be removed.

JyotinderSingh replied:
This will cause the server to crash, instead let's return an UnimplementedError for such commands.

pratikpandey21 replied:
Okay cool.

JyotinderSingh commented on main.go:
```diff
@@ -51,23 +62,73 @@ func main() {
+	go func() {
+		defer serverWg.Done()
+		// Run the HTTP server
+		err := httpServer.Run(ctx)
```
Should we create a cancellable context from the ctx for the `httpServer` and `asyncServer` to maintain consistency with the shard manager? Or change the shard manager to use the common ctx.

(Actually I would like to know what the preferred practice is)

pratikpandey21 replied:
So, we create a child context when we need to maintain a separate lifecycle, i.e we can cancel the child context without impacting the parent.
Any issues with the Async or HTTP server would mean we want to shut things down and cancel all child contexts. I don't think that we would like to keep Shard Manager at the common context level.

We can, however, create separate contexts for each of them, but we need to remember that each cancellable context spawns a goroutine to handle cancellation. We're just talking about two more, so it should be fine.

JyotinderSingh commented on internal/server/utils/redisCmdAdapter.go:
```diff
@@ -0,0 +1,72 @@
+
+			// Define keys to exclude and process their values first
+			// Update as we support more commands
+			priorityKeys := []string{"key", "field", "path", "value"}
```
Let's use a fixed size array here instead of a slice.

JyotinderSingh commented on internal/server/utils/redisCmdAdapter.go:
```diff
@@ -0,0 +1,72 @@
+)
+
+func ParseHTTPRequest(r *http.Request) (*cmd.RedisCmd, error) {
+	prefix := strings.TrimPrefix(r.URL.Path, "/")
```
nit: just for better clarity.
```suggestion
	command := strings.TrimPrefix(r.URL.Path, "/")
```

JyotinderSingh commented on internal/server/utils/redisCmdAdapter.go:
```diff
@@ -0,0 +1,72 @@
+			}
+
+			// Process remaining keys
+			for k, v := range jsonBody {
+				// Handle unary operations like 'nx' where value is "true"
+				if valStr, ok := v.(string); ok && valStr == "true" {
+					args = append(args, k)
+				} else {
+					switch val := v.(type) {
+					case map[string]interface{}, []interface{}:
+						jsonValue, err := json.Marshal(val)
+						if err != nil {
+							return nil, err
+						}
+						args = append(args, string(jsonValue))
+					default:
+						args = append(args, fmt.Sprintf("%v", val))
+					}
+				}
+			}
```
Maybe this can be made clearer in this fashion:

```suggestion
	    // Process remaining keys in the JSON body
            for key, val := range jsonBody {
                switch v := val.(type) {
                case string:
                    // Handle unary operations like 'nx' where value is "true"
                    if v == "true" {
                        args = append(args, key)
                    } else {
                        args = append(args, v)
                    }
                case map[string]interface{}, []interface{}:
                    // Marshal nested JSON structures back into a string
                    jsonValue, err := json.Marshal(v)
                    if err != nil {
                        return nil, err
                    }
                    args = append(args, string(jsonValue))
                default:
                    // Append other types as strings
                    args = append(args, fmt.Sprintf("%v", v))
                }
            }
```

JyotinderSingh reviewed (changes requested):
Thanks for adding support for HTTP @pratikpandey21! The changes look good, have left a few small comments.

JyotinderSingh commented:
There seem to be a few conflicts as well, please resolve them too.

JyotinderSingh reviewed (approved):
Thanks for addressing the reviews @pratikpandey21! The changes look good, LGTM.

There is one linter error, please fix it and I will proceed with the merge.

pratikpandey21 commented:
> Thanks for addressing the reviews @pratikpandey21! The changes look good, LGTM.
> 
> There is one linter error, please fix it and I will proceed with the merge.

Fixed. Thanks @JyotinderSingh

nithinkm8055 commented on internal/server/httpServer.go:
```diff
@@ -0,0 +1,142 @@
+	var wg sync.WaitGroup
+	var err error
+
+	_, cancelHTTP := context.WithCancel(ctx)
```
The cancellable context is not used. It can be removed.

nithinkm8055 commented on internal/server/httpServer.go:
```diff
@@ -0,0 +1,142 @@
+	go func() {
+		defer wg.Done()
+		log.Infof("HTTP Server running on Port%s", s.httpServer.Addr)
+		err = s.httpServer.ListenAndServe()
```
Is there a reason to not log this error.
